designs/0030-complex-numbers.md

```diff
@@ -0,0 +1,355 @@
+
+### Matrix and array constructors
+
+The array and matrix constructors will extend to complex numbers. Like their behavior with real numbers, the result should be the promoted type.  For example,
+
+```
+complex z1;
+complex z2;
+complex_vector[2] v = [z1, z2]';
+```
+
+is fine, as is
+
+```
+complex z1;
+real x2;
+int n3;
+complex_vector[3] v = [z1, x2, n3]';
+```
+
+but not
+
+```
+complex z1;
+real x2;
+vector[2] v = [z1, x2]';  // ILLEGAL, [z1, x2]' type is complex_vector
+```
+
+Complex matrices should be constructible from sequences of complex row vectors,
+as in
+
+```
+complex z;
+complex_matrix[2, 2] m
+  = [[1, z], [z, 1]];
+```
+
+Arrays are constructed as usual,
+
+```
+complex z1;  complex z2;
+complex z[5] = {z1, z2, z3, 1, 2.8};
+```
+
+As with vectors, promotion to the richest type found for an element will be carried out for array, vector, and matrix construction.
+
+
+### Elementwise operators
+
+The elementwise vector and matrix operations, `.*` and  `./`, should work as expected, carrying out their operations elementwise.    They should allow mixed complex and real results, e.g.,
+
+```
+vector[2] x;
+complex_vector[2] y;
+complex_vector[2] z = x .* y;
+```
+
+### Scalar functions
+
+Every scalar complex function available in the C++ standard library `<complex>` header will be included.  These are already implemented in the Stan math library.
+
+### Matrix functions
+
+There are no matrix functions exposed through the math library, but fast Fourier transforms (FFT), asymmetric eigendecomposition, and Schur decomposition have all been implemented and are part of the unit tests for complex numbers.
+
+
+## C++ model class
+
+Stan programs are converted to C++ class definitions by the stanc3 transpiler.   There are two components in this compiled class that.  This section covers data input and parameter output.  User-facing I/O is handled by the interfaces.
+
+### Data input format
+
+The C++ model class uses an object of class `var_context` to input variables declared in the data block.  This object needs to be extended to read complex number objects.  To read a complex number, we must read a sequence of two real numbers to pass to the C++ `complex<double>` constructor.
+
+### Parameter output format
+
+Parameters are output as an array.  Complex numbers will be sequenced with their real components before their imaginary components.  The output names for a complex variable `z` will be `z.r` and `z.i` for the real and imaginary components.
```

Comment:
These ended up as `z.real` and `z.imag`, not `r` and `i`

Reply:
At least it's only one line :-).  The doc should change.